### PR TITLE
[reference-binding] Add _consuming/_mutating/_borrowing as var decl bindings

### PR DESCRIPTION
**Pull request by gottesmm**
Title says it all.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 1]
```diff
@@ -1179,6 +1179,9 @@ public let DECL_NODES: [Node] = [
           .keyword(text: "let"),
           .keyword(text: "func"),
           .keyword(text: "inout"),
+          .keyword(text: "mutating"),
+          .keyword(text: "consuming"),
+          .keyword(text: "borrowing"),
```
Comment (ahoppen) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:1184]: Is `import mutating Module.var` something we want to support? Same for `consuming`. Could you also add test cases for it?

File: CodeGeneration/Sources/SyntaxSupport/StmtNodes.swift [hunk 2]
```diff
@@ -440,7 +440,10 @@ public let STMT_NODES: [Node] = [
       Child(
         name: "BindingSpecifier",
         deprecatedName: "BindingKeyword",
-        kind: .token(choices: [.keyword(text: "let"), .keyword(text: "var"), .keyword(text: "inout")])
+        kind: .token(choices: [
+          .keyword(text: "let"), .keyword(text: "var"), .keyword(text: "inout"),
+          .keyword(text: "mutating"), .keyword(text: "borrowing"), .keyword(text: "consuming"),
+        ])
```
Comment (ahoppen) [inline CodeGeneration/Sources/SyntaxSupport/StmtNodes.swift:446]: AFAICT we don’t have tests for these. Could you add some? Should be `if mutating x = x {}`

File: Sources/SwiftParser/TokenSpecSet.swift [hunk 3]
```diff
@@ -405,7 +405,7 @@ enum DeclarationModifier: TokenSpecSet {
     case .internal: return .keyword(.internal)
     case .isolated: return TokenSpec(.isolated, recoveryPrecedence: .declKeyword)
     case .lazy: return .keyword(.lazy)
-    case .mutating: return .keyword(.mutating)
+    case .mutating: return TokenSpec(.mutating, recoveryPrecedence: .declKeyword)
```
Comment (ahoppen) [inline Sources/SwiftParser/TokenSpecSet.swift:408]: `mutating` already has a `TokenPrecedence` of `declKeyword`, so I don’t think this change is needed.

https://github.com/apple/swift-syntax/blob/250b431726c7b05309770caaa9989a4166ea6ba7/Sources/SwiftParser/TokenPrecedence.swift#L229-L233

File: Sources/SwiftParser/Patterns.swift [hunk 4]
```diff
@@ -93,7 +102,10 @@ extension Parser {
       )
     case (.let, let handle)?,
       (.var, let handle)?,
-      (.inout, let handle)?:
+      (.inout, let handle)?,
+      (.mutating, let handle)?,
+      (.borrowing, let handle)?,
+      (.consuming, let handle)?:
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:108]: You don’t need to add all the cases to `ExpectedTokens` here, if you 
- Remove `let`, `var`, `inout` from `ExpectedTokens`
- Add a new `switch self.at(anyIn: ValueBindingPatternSyntax.BindingSpecifierOptions.self) {` here.

`ValueBindingPatternSyntax.BindingSpecifierOptions` is generated from the options you specify in `CodeGeneration`. It’s a recent-ish feature, that’s why it isn’t used yet. But a great opportunity to clean up the codebase a little.

File: Sources/SwiftParser/Patterns.swift [hunk 5]
```diff
@@ -217,7 +229,10 @@ extension Parser {
     switch self.at(anyIn: MatchingPatternStart.self) {
     case (.var, let handle)?,
       (.let, let handle)?,
-      (.inout, let handle)?:
+      (.inout, let handle)?,
+      (.mutating, let handle)?,
+      (.borrowing, let handle)?,
+      (.consuming, let handle)?:
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:235]: Same comment of using `ValueBindingPatternSyntax.BindingSpecifierOptions` also applies here.

File: Sources/SwiftParser/Patterns.swift [hunk 6]
```diff
@@ -274,6 +289,9 @@ extension Parser.Lookahead {
       case `var`
       case leftParen
       case `inout`
+      case `mutating`
+      case borrowing
+      case consuming
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:294]: Same comment of using `ValueBindingPatternSyntax.BindingSpecifierOptions` applies here.

**Review by ahoppen — COMMENTED** [review]
Mostly looks good to me. I’ve got one suggestion to simplify the `TokenSpecSets` by using the auto-generated ones.

Also, since you are touching the test cases, would you mind splitting them so that each statement/case/variable its own `assertParse` call? It just makes it easier to debug the tests in case one of them failed because you start off with less source and you don’t actually need to reduce the input.

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift (hunk 1 above)
Comment (gottesmm) [reply CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:1184]: @ahoppen are there any tests like this that exist already? I did a git grep for import var/let/etc.

Comment (ahoppen) [reply CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:1184]: I don’t think so. DeclarationTests.swift seems like a good place to add them though

**Comment by gottesmm** [comment]
@swift-ci test

File: Sources/SwiftParser/TokenSpecSet.swift [hunk 7]
```diff
@@ -21,6 +21,38 @@ protocol TokenSpecSet: CaseIterable {
...
+
+  init?(lexeme: Lexer.Lexeme) {
+    if let x = LHS.init(lexeme: lexeme) {
+      self = .lhs(x)
+      return
+    }
+    if let y = RHS.init(lexeme: lexeme) {
+      self = .rhs(y)
+      return
+    }
```
Comment (ahoppen) [inline Sources/SwiftParser/TokenSpecSet.swift:37]: Nit: You don’t need `.init`

```suggestion
    if let x = LHS(lexeme: lexeme) {
      self = .lhs(x)
      return
    }
    if let y = RHS(lexeme: lexeme) {
      self = .rhs(y)
      return
    }
```

File: Sources/SwiftParser/TokenSpecSet.swift [hunk 8]
```diff
@@ -227,7 +259,10 @@ enum ContextualDeclKeyword: TokenSpecSet {
   }
 }
 
-enum DeclarationKeyword: TokenSpecSet {
+/// DeclarationKeywords that are only semantic DeclarationKeywords.
+///
+/// Excludes things like ValueBindingPatterns. They are injected into DeclarationKeyword via an either.
```
Comment (ahoppen) [inline Sources/SwiftParser/TokenSpecSet.swift:264]:
```suggestion
/// Excludes things like ValueBindingPatterns. They are injected into `DeclarationKeyword` via an `EitherTokenSpecSet`
```

File: Sources/SwiftParser/TokenSpecSet.swift [hunk 9]
```diff
@@ -719,32 +750,31 @@ enum ExpressionPrefixOperator: TokenSpecSet {
   }
 }
 
-enum MatchingPatternStart: TokenSpecSet {
+/// MatchingPatternStart without ValueBindingPatterns.
+///
+/// We use an EitherTokenSpecSet to inject ValueBindingPatterns.
```
Comment (ahoppen) [inline Sources/SwiftParser/TokenSpecSet.swift:755]:
```suggestion
/// `MatchingPatternStart` without `ValueBindingPatterns`.
///
/// We use an `EitherTokenSpecSet` to inject `ValueBindingPatterns`.
```

**Review by ahoppen — COMMENTED** [review]
Thanks for adding `EitherTokenSpecSet`. I think it will start to be useful in other cases as well.

Looks good to me, I just think that we should be guarding the feature behind `perimentalFeatures.contains(.whateverYouWantToCallThisFeature)`.

File: Sources/SwiftParser/TokenSpecSet.swift [hunk 10]
```diff
@@ -227,7 +259,10 @@ enum ContextualDeclKeyword: TokenSpecSet {
   }
 }
 
-enum DeclarationKeyword: TokenSpecSet {
+/// DeclarationKeywords that are only semantic DeclarationKeywords.
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply Sources/SwiftParser/TokenSpecSet.swift:262]: What I am trying to say here is that they are not ValueBinding ones. I will fix that.

**Comment by gottesmm** [comment]
@swift-ci test

File: Sources/SwiftParser/Declarations.swift [hunk 11]
```diff
@@ -210,81 +210,99 @@ extension Parser {
     // while recovering to the declaration start.
     let recoveryPrecedence = inMemberDeclList ? TokenPrecedence.closingBrace : nil
 
-    switch self.canRecoverTo(anyIn: DeclarationKeyword.self, overrideRecoveryPrecedence: recoveryPrecedence) {
-    case (.import, let handle)?:
+    CanRecoverToSwitch: switch self.canRecoverTo(anyIn: DeclarationKeyword.self, overrideRecoveryPrecedence: recoveryPrecedence) {
```
Comment (ahoppen) [inline Sources/SwiftParser/Declarations.swift:213]: I don’t actually think we need the label. If we `break` in the nested `switch`, then it just executes to the end of the outer `switch`, which is the same as `break CanRecoverToSwitch`.

---

If we do need it: We currently all caps our labels, so for consistency, this should be `CAN_REVOVER_TO_SWITCH`.

File: Sources/SwiftParser/Declarations.swift [hunk 12]
```diff
@@ -210,81 +210,99 @@ extension Parser {
...
-    case (.pound, let handle)?:
+    case (.lhs(.pound), let handle)?:
       return RawDeclSyntax(self.parseMacroExpansionDeclaration(attrs, handle))
+    case (.rhs(let binding), let handle)?:
+      switch binding {
+      case ._mutating:
+        fallthrough
+      case ._borrowing:
+        fallthrough
+      case ._consuming:
```
Comment (ahoppen) [inline Sources/SwiftParser/Declarations.swift:256]: Slightly shorter version

```suggestion
      case ._mutating, ._borrowing, ._consuming:
```

File: Sources/SwiftParser/Declarations.swift [hunk 13]
```diff
@@ -210,81 +210,99 @@ extension Parser {
...
+        guard experimentalFeatures.contains(.referenceBindings) else {
+          break CanRecoverToSwitch
+        }
+        fallthrough
+      case .let:
+        fallthrough
+      case .var:
+        fallthrough
+      case .inout:
```
Comment (ahoppen) [inline Sources/SwiftParser/Declarations.swift:265]: Slightly shorter version

```suggestion
      case .let, .var, .inout:
```

File: Sources/SwiftParser/Patterns.swift [hunk 14]
```diff
@@ -91,32 +86,48 @@ extension Parser {
...
-      if self.currentToken.isLexerClassifiedKeyword, !self.atStartOfLine {
-        // Recover if a keyword was used instead of an identifier
-        let keyword = self.consumeAnyToken()
+    case (.rhs(let binding), let handle)?:
+      switch binding {
+      case ._mutating:
+        fallthrough
+      case ._borrowing:
+        fallthrough
+      case ._consuming:
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:95]:
```suggestion
      case ._mutating, ._borrowing, ._consuming:
```

File: Sources/SwiftParser/Patterns.swift [hunk 15]
```diff
@@ -91,32 +86,48 @@ extension Parser {
...
+        guard experimentalFeatures.contains(.referenceBindings) else {
+          break
+        }
+        fallthrough
+      case .let:
+        fallthrough
+      case .var:
+        fallthrough
+      case .inout:
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:104]:
```suggestion
      case .let, .var, .inout:
```

File: Sources/SwiftParser/Patterns.swift [hunk 16]
```diff
@@ -214,20 +225,8 @@ extension Parser {
...
-      let value = self.parseMatchingPattern(context: .bindingIntroducer)
-      return RawPatternSyntax(
-        RawValueBindingPatternSyntax(
-          bindingSpecifier: bindingSpecifier,
-          pattern: value,
-          arena: self.arena
-        )
-      )
-    case (.is, let handle)?:
+    OuterSwitch: switch self.at(anyIn: MatchingPatternStart.self) {
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:228]: I don’t actually think we need the label. If we `break` in the nested `switch`, then it just executes to the end of the outer `switch`, which is the same as `break OuterSwitch`.

---

If we do need it: We currently all caps our labels, so for consistency, this should be `OUTER_SWITCH`.

File: Sources/SwiftParser/Patterns.swift [hunk 17]
```diff
@@ -237,23 +236,52 @@ extension Parser {
...
-        // FIXME: This is pretty gross. Let's find a way to disambiguate let
-        // binding patterns much earlier.
-        return RawPatternSyntax(pat.pattern)
+    case (.rhs(let binding), let handle)?:
+      switch binding {
+      case ._mutating:
+        fallthrough
+      case ._borrowing:
+        fallthrough
+      case ._consuming:
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:245]:
```suggestion
      case ._mutating, ._borrowing, ._consuming:
```

File: Sources/SwiftParser/Patterns.swift [hunk 18]
```diff
@@ -237,23 +236,52 @@ extension Parser {
...
+        guard experimentalFeatures.contains(.referenceBindings) else {
+          break OuterSwitch
+        }
+        fallthrough
+      case .let:
+        fallthrough
+      case .var:
+        fallthrough
+      case .inout:
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:254]:
```suggestion
      case .let, .var, .inout:
```

File: Sources/SwiftParser/Declarations.swift [hunk 19]
```diff
@@ -210,81 +210,99 @@ extension Parser {
...
+        // Handle all cases of `ValueBindingPatternSyntax.BindingSpecifierOptions` in the same way.
```
Comment (ahoppen) [inline Sources/SwiftParser/Declarations.swift:266]: I don’t think we need this comment anymore if we exhaustively switch over `rhs`.

File: Sources/SwiftParser/Patterns.swift [hunk 20]
```diff
@@ -91,32 +86,48 @@ extension Parser {
...
+        guard experimentalFeatures.contains(.referenceBindings) else {
+          break
+        }
+        fallthrough
+      case .let:
+        fallthrough
+      case .var:
+        fallthrough
+      case .inout:
+        // Handle all cases of `ValueBindingPatternSyntax.BindingSpecifierOptions` in the same way.
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:105]: I don’t think we need this comment anymore if we exhaustively switch over `rhs`.

File: Sources/SwiftParser/Patterns.swift [hunk 21]
```diff
@@ -291,25 +313,40 @@ extension Parser.Lookahead {
...
+      case ._consuming:
+        guard experimentalFeatures.contains(.referenceBindings) else {
+          return false
+        }
+        fallthrough
+      case .let:
+        fallthrough
+      case .var:
+        fallthrough
+      case .inout:
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:345]:
```suggestion
      case .let, .var, .inout:
```

File: Sources/SwiftParser/Patterns.swift [hunk 22]
```diff
@@ -291,25 +313,40 @@ extension Parser.Lookahead {
...
-      self.eat(handle)
-      return self.canParsePattern()
-    case (.leftParen, _)?:
+    case (.rhs(let binding), let handle)?:
+      switch binding {
+      case ._mutating:
+        fallthrough
+      case ._borrowing:
+        fallthrough
+      case ._consuming:
```
Comment (ahoppen) [inline Sources/SwiftParser/Patterns.swift:336]:
```suggestion
      case ._mutating, ._borrowing, ._consuming:
```

**Review by ahoppen — APPROVED** [review]
Looks great to me. Just a couple syntactic improvement suggestions. If you want to land this now, you can do them in a follow-up PR.

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**